src/System.Private.CoreLib/shared/System/SpanHelpers.Byte.cs

```diff
@@ -255,29 +268,124 @@ public static unsafe int IndexOf(ref byte searchSpace, byte value, int length)
                 index += 1;
             }
 
-            if (Vector.IsHardwareAccelerated && ((int)(byte*)index < length))
+            if (Avx2.IsSupported)
             {
-                nLength = (IntPtr)((length - (int)(byte*)index) & ~(Vector<byte>.Count - 1));
+                if ((int)(byte*)index < length)
+                {
+                    nLength = (IntPtr)((length - (int)(byte*)index) & ~(Vector256<byte>.Count - 1));
+                    if ((byte*)nLength > (byte*)index)
+                    {
+                        Vector256<byte> comparison = Vector256.Create(value);
+                        do
+                        {
+                            Vector256<byte> search = Unsafe.ReadUnaligned<Vector256<byte>>(ref Unsafe.AddByteOffset(ref searchSpace, index));
+                            int matches = Avx2.MoveMask(Avx2.CompareEqual(comparison, search));
+                            if (matches == 0)
+                            {
+                                index += Vector256<byte>.Count;
+                                continue;
+                            }
+                            // Find offset of first match
+                            else if (Bmi1.IsSupported)
+                            {
+                                return ((int)(byte*)index) + (int)Bmi1.TrailingZeroCount((uint)matches);
```

Comment:
Aside: would it make sense to implement `TrailingZeroCountFallback` as software-fallback for the intrinsic `Bmi1.TrailingZeroCount`?

Reply:
`Bmi1.TrailingZeroCount` has no software fallback; however you are probably asking for this: https://github.com/dotnet/corefx/issues/32269?